src/coreclr/vm/gchelpers.cpp

```diff
@@ -855,8 +855,7 @@ STRINGREF AllocateString( DWORD cchStringLength )
 
     SetTypeHandleOnThreadForAlloc(TypeHandle(g_pStringClass));
 
-    GC_ALLOC_FLAGS flags = GC_ALLOC_NO_FLAGS;
-    if (totalSize >= g_pConfig->GetGCLOHThreshold())
+    if (totalSize >= g_pConfig->GetGCLOHThreshold() && !(flags & GC_ALLOC_PINNED_OBJECT_HEAP))
```

Comment:
This should match what `AllocateSzArray` does. It just sets both `GC_ALLOC_PINNED_OBJECT_HEAP` and `GC_ALLOC_LARGE_OBJECT_HEAP` flags for this case.